### ISSUE TITLE
Change default download repo 

### ISSUE THREAD

**Issue by Kopo942** [slimjar/slimjar#52]
Hey, is there a way to change the repository list for downloading at runtime? Slimjar currently uses your own repo (repo.vshnv.tech) to download many dependencies for us, however we'd like to rely on Maven central by default since they use a CDN & are more reliable in general.

**Comment by Vshnv** [comment]
You can use ApplicationBuilder#internalRepositories.

Altho it is generally discouraged to use central directly for this purpose as all users downloading from central would take up more of their bandwidth, you could try hosting your own mirror.

**Comment by Kopo942** [comment]
To be honest, bandwidth really shouldn't be an issue for a central repository using a global CDN (Fastly). I also wouldn't like to rely on a 3rd-party repository for reliability: if the repo is down, our plugin might fail to start.

Hosting our own mirror isn't really feasible budget-wise.

**Comment by Vshnv** [comment]
Was just reiterating what was in the maven docs

**Comment by AuroraLS3** [comment]
Using ApplicationBuilder#internalRepositories did the trick. Thanks!

**Comment by AuroraLS3** [comment]
Hi, apparently the internal repositories method call does not stop calls to other repositories - what should I do? https://github.com/plan-player-analytics/Plan/issues/2058

**Comment by Vshnv** [comment]
Looking into this issue, will get back to you asap

**Comment by AuroraLS3** [comment]
Hi, how's it going?

**Comment by AuroraLS3** [comment]
I tried to analyze the data flow but I ran into some trouble since the cache cuts off IntelliJ data flow analyzer.
Here's what I figured if it helps

- `ApplicationBuilder` users `SimpleDependencyInjectorFactory` and then `SimpleDependencyInjector`
- `SimpleDependencyInjector` creates an `InjectionHelper` with the paramters from `ApplicationBuilder`
- `InjectionHelper` downloads dependencies using `DependencyDownloader`
- `InjectionHelpreFactory` creates it with 
   ```java
        final DependencyResolver resolver = resolverFactory.create(repositories, preResolvedResults, enquirerFactory);
        final DependencyDownloader downloader = downloaderFactory.create(outputWriterFactory, resolver, verifier.create(resolver));
    ```
- `repositories` comes from `DependencyData`
    ```java
        public InjectionHelper create(final DependencyData data, final Map<String, ResolutionResult> preResolvedResults) throws IOException, NoSuchAlgorithmException, URISyntaxException {
        final Collection<Repository> repositories = mirrorSelector
                .select(data.getRepositories(), data.getMirrors());
    ```
- Running data flow analyzer on `DependencyData data` has one constructed with
   ```java
       new DependencyData(
                    Collections.emptySet(),
                    Collections.emptySet(),
                    Collections.emptySet(),
                    Collections.emptySet()
            );
   ```
   and others seem to come from cache by reading json. I don't unfortunately have enough knowledge about the project to continue the data flow analysis further from here. My best guess is the repositories given to mirrorSelector gets the default mirrors added in somewhere

**Comment by Vshnv** [comment]
Hmm, what I was assuming is that this is due to the pre-resolution which doesnt not really have access to the ones declared in ApplicationBuilder#internalRepositories so it uses the default mirror for the pre-resolution stage.

**Comment by AuroraLS3** [comment]
That would also make sense

**Comment by Kopo942** [comment]
Hey @Vshnv, any ETA on this getting fixed?

**Comment by Vshnv** [comment]
I'll try to get it fixed in a day or two

**Comment by Vshnv** [comment]
Looking into this a bit, it shouldnt be using my repo as a mirror for compile-time resolution due to this [commit](https://github.com/slimjar/slimjar/commit/f655afcdca63d093ce2190cf6fae13ec9d83449f#diff-e783a61cfe7d7c5a9228328a7417990ab2155d743e07d053538e419cd98c6238) `f655afcdca63d093ce2190cf6fae13ec9d83449f`. 

What I can see from that is the issue is the resolver still having my repo in the all repos collection which makes it resolve a few dependencies on that. I'll try to figure out a way to overcome this.

**Comment by Vshnv** [comment]
Switching this to a priority queue instead might work I guess (https://github.com/slimjar/slimjar/blob/master/slimjar/src/main/java/io/github/slimjar/resolver/CachingDependencyResolver.java#L45).

**Comment by Kopo942** [comment]
I believe a PriorityQueue expects a class that `implements Comparable<T>`, so changing
```java
.internalRepositories(Collections.singletonList(new Repository(new URL("https://papermc.io/repo/repository/maven-public/"))))
```
to
```java
Queue<Repository> repositories = new PriorityQueue<>();
repositories.add(new Repository(new URL("https://papermc.io/repo/repository/maven-public/")));
// snip
.internalRepositories(repositories)
```
would throw a ClassCastException since `Repository` doesn't implement `Comparable`.

**Comment by Vshnv** [comment]
Isnt there an alternative option of simply providing a comparator as the second parameter?

**Comment by AuroraLS3** [comment]
Yes you can give a comparator as a paremeter to PriorityQueue constructor

**Comment by Vshnv** [comment]
Alrighty, ill give that a shot to make sure central takes the highest priority in the collection

**Comment by AuroraLS3** [comment]
Hi, what is the status on this? :) 
I'm planning next release of my project in 2 weeks

Anything I can do to help?
